spec/optional_query_cache_spec.rb

```diff
@@ -0,0 +1,65 @@
+require 'action_controller/metal/exceptions'
+require 'active_record'
+require 'rack'
+
+require 'spec_helper'
+
+require 'optional_query_cache'
+
+RSpec.describe OptionalQueryCache::QueryCache do
+  let(:app) { ->(env) { [200, env, 'app'] } }
+
+  let :subject do
+    described_class.new(app)
+  end
+
+  let :env do
+    double.tap do |env|
+      allow(env).to receive(:[]).with('REQUEST_PATH').and_return('')
+    end
+  end
+
+  it 'permits query caching by default' do
+    expect(ActiveRecord::Base).to receive(:connection).and_return(
+      double(
+        'connection',
+        query_cache_enabled: true,
+        enable_query_cache!: nil
+      )
+    )
+    expect(env).to receive(:[]).with('action_dispatch.routes').and_return(
+      double(
+        'routes',
+        recognize_path: {}
+      )
+    )
+    subject.call(env)
+  end
+
+  it 'disables query caching for specified routes' do
+    expect(ActiveRecord::Base).not_to receive(:connection)
+    expect(env).to receive(:[]).with('action_dispatch.routes').and_return(
+      double(
+        'routes',
+        recognize_path: {
+          disable_query_cache: true
+        }
+      )
+    )
+    subject.call(env)
+  end
+
+  it 'permits query caching for unrecognised routes' do
```

Comment:
when you've got two or more very similar examples which test the same thing with different parameters, its often clearer to structure it with context blocks, like:

```ruby
  context "when the route is unrecognised" do
    # set up context that emulates the route being unrecognized, e.g.
    before do
      allow(env).to receive_message_chain(:[],:recognize_path).and_raise(
        ActionController::RoutingError.new('Unrecognised route')
      )
    end
    it "permits query caching" do
      expect(ActiveRecord::Base).to receive(:connection).and_return(double('mock connection'))
    end
  end

  context "when the route is specified" do
    before do
      allow(env).to receive(:[]).with('action_dispatch.routes').and_return(
        double(
          'routes',
          recognize_path: {
            disable_query_cache: true
          }
        )
      )
    end

    it "disables query caching" do
      expect(ActiveRecord::Base).not_to receive(:connection)
    end
  end
```